sdk/python/lib/pulumi/runtime/resource.py

```diff
@@ -147,6 +147,28 @@ def do_rpc_call():
 
     asyncio.ensure_future(RPC_MANAGER.do_rpc("register resource", do_register)())
 
-def register_resource_outputs(_res: 'Resource', _inputs: 'Union[Inputs, Awaitable[Inputs], Output[Inputs]]'):
-    # TODO(swgillespie) needed for stack outputs and component resources.
-    pass
+def register_resource_outputs(res: 'Resource', outputs: 'Union[Inputs, Awaitable[Inputs], Output[Inputs]]'):
+    async def do_register_resource_outputs():
+        urn = await res.urn.future()
+        serialized_props = await rpc.serialize_properties(outputs, [])
+        log.debug(f"register resource outputs prepared: urn={urn}, props={serialized_props}")
+        monitor = settings.get_monitor()
+        req = resource_pb2.RegisterResourceOutputsRequest(urn=urn, outputs=serialized_props)
+
+        def do_rpc_call():
+            try:
+                return monitor.RegisterResourceOutputs(req)
+            except grpc.RpcError as exn:
+                # See the comment on invoke for the justification for disabling
+                # this warning
+                # pylint: disable=no-member
+                if exn.code() == grpc.StatusCode.UNAVAILABLE:
+                    sys.exit(0)
```

Comment:
Hm, interesting. Can we guarantee that there's no more user code that might run?

Reply:
That is to say, if we threw an exception and continued, would it be possible for more user code to run?

Reply:
> That is to say, if we threw an exception and continued, would it be possible for more user code to run?

It would be dangerous for us to do so, because 1) if a user caught the exception and handled it somehow, they would have a partially-initialized Resource object whose output properties would never be resolved. Any `apply` on them would deadlock and all future resource operations using those outputs would also deadlock waiting for resolution. In my eyes, it's much safer in this case to exit.

Reply:
Okay. We should document this, though I'm not sure where.

> they would have a partially-initialized Resource object whose output properties would never be resolved. Any apply on them would deadlock and all future resource operations using those outputs would also deadlock waiting for resolution. In my eyes, it's much safer in this case to exit.

This is a very good point. If we ever change our minds re: promptly exiting the process, we could explore resolving these properties s.t. any `apply` throws.